File: leetcode/C++/264.cpp
#include <vector>

using namespace std;

class Solution
{
  public:
    int ids[3]{0};
    vector<int> uglys;

    int nthUglyNumber(int n)
    {
        uglys.resize(n);
        uglys[0] = 1;
        for (int i = 1; i < n; ++i)
        {
            int a = uglys[ids[0]] * 2;
            int b = uglys[ids[1]] * 3;
            int c = uglys[ids[2]] * 5;
            int min = minInt(a, minInt(b, c));
            if (a == min)
                ++ids[0];
            if (b == min)
                ++ids[1];
            if (c == min)
                ++ids[2];
            uglys[i] = min;
        }
        return uglys[n - 1];
    }
    int minInt(int a, int b)
    {
        return a < b ? a : b;
    }
};

int main()
{
    Solution s;
    int res = s.nthUglyNumber(1695);
    return 0;
}

File: leetcode/Go/264.go
package main

func nthUglyNumber(n int) int {
	var num [1695]int
	num[0] = 1
	var x, y, z = 0, 0, 0
	for i := 1; i < n; i++ {
		//始终确保 a, b, c 都大于 num[i-1]
		var a = num[x] * 2
		var b = num[y] * 3
		var c = num[z] * 5
		var min = minInt(a, minInt(b, c))
		if a == min {
			x++
		}
		if b == min {
			y++
		}
		if c == min {
			z++
		}
		num[i] = min
	}
	return num[n-1]
}
